File: submissions/Chimoney-Python/pychimoney/Account.py
from pychimoney import BaseAPI


class Account(BaseAPI):
    """
    Account Endpoints Wrapper

    This class wraps the Account endpoints of the Chi Money API.

    list of endpoints:
        - transactions_by_issue_id
        - all_transaction
        - transaction_by_id
        - account_transfer
        - delete_unpaid_transaction
    """

    def trasactions_by_issue_id(self, issue_id, sub_account=None) -> dict:
        """
        This function returns a list of transactions by issue ID.

        :param issue_id: The issue ID of the transaction.
        :type issue_id: str
        :param sub_account: The sub account of the transaction.
        :type sub_account: str
        :return: The response from the Chi Money API.
        :rtype: dict
        """
        if not isinstance(issue_id, str):
            raise TypeError("Issue ID must be a string.")

        if not issue_id:
            raise ValueError("Issue ID is required.")

        params = {"issueId": issue_id}
        payload = {}
        if sub_account:
            payload["subAccount"] = sub_account

        return self._handle_request(
            "POST", "/v0.2/accounts/issue-id-transactions", params=params, data=payload
        )

    def all_transaction(self, sub_account=None) -> dict:
        """
        This function returns a list of transactions by account.

        :param sub_account: The sub account of the transaction.
        :type sub_account: str
        :return: The response from the Chi Money API.
        :rtype: dict
        """
        payload = {}
        if sub_account:
            payload["subAccount"] = sub_account

        return self._handle_request(
            "POST",
            "/v0.2/accounts/transactions",
            data=payload,
        )

    def transaction_by_id(self, transaction_id, sub_account=None) -> dict:
        """
        This function returns a transaction by ID.

        :param transaction_id: The ID of the transaction.
        :type transaction_id: str
        :param sub_account: The sub account of the transaction.
        :type sub_account: str
        :return: The response from the Chi Money API.
        :rtype: dict
        """
        if not isinstance(transaction_id, str):
            raise TypeError("Transaction ID must be a string.")

        if not transaction_id:
            raise ValueError("Transaction ID is required.")

        params = {"id": transaction_id}
        payload = {}
        if sub_account:
            payload["subAccount"] = sub_account

        return self._handle_request(
            "GET", "/v0.2/accounts/transaction", params=params, data=payload
        )

    def account_transfer(self, receiver, amount, wallet, sub_account=None) -> dict:
        """
        This function transfers funds from one account to another.

        :param receiver: The receiver of the funds.
        :type receiver: str
        :param amount: The amount of the funds.
        :type amount: float
        :param wallet: The wallet of the funds.
        :type wallet: str
        :param sub_account: The sub account of the transaction.
        :type sub_account: str
        :return: The response from the Chi Money API.
        :rtype: dict
        """
        if not isinstance(reciver, str):
            raise TypeError("Reciver must be a string.")

        if not isinstance(amount, int):
            raise TypeError("Amount must be an integer.")

        if not isinstance(wallet, str):
            raise TypeError("Wallet must be a string.")

        if not receiver and amount and wallet:
            raise ValueError("Reciver, amount and wallet are required.")

        payload = {
            "receiver": receiver,
            "amount": amount,
            "wallet": wallet,
        }

        if sub_account:
            payload["subAccount"] = sub_account

        return self._handle_request(
            "POST",
            "/v0.2/accounts/transfer",
            data=payload,
        )

    def delete_unpaid_transaction(self, transaction_id, sub_account=None):
        """
        This function deletes an unpaid transaction.

        :param transaction_id: The ID of the transaction.
        :type transaction_id: str
        :param sub_account: The sub account of the transaction.
        :type sub_account: str
        :return: The response from the Chi Money API.
        :rtype: dict
        """
        if not isinstance(transaction_id, str):
            raise TypeError("Transaction ID must be a string.")

        if not transaction_id:
            raise ValueError("Transaction ID is required.")

        params = {"chiRef": transaction_id}

        if sub_account:
            params["subAccount"] = sub_account
        return self._handle_request(
            "DELETE", "/v0.2/accounts/delete-unpaid-transaction", params=params
        )

File: submissions/chimoney-js/modules/Account.js
const Joi = require("joi");
const { ValueError, TypeError } = require("../Errors");
const {
  handleRequest,
  HTTPMETHODS,
  formatJoiErrors,
} = require("../utils/helpers");

/**
 * This function gets all transactions by IssueID
 * @param {string} issueID The issueID of the transaction
 * @param {string?} subAccount The subAccount of the transaction
 * @returns The response from the Chi Money API
 */
async function getTransactionsByIssueID(issueID, subAccount = null) {
  if (!issueID) throw new ValueError("issueId is required");

  if (typeof issueID !== "string")
    throw new TypeError("issueId must be a string");

  const payload = {};
  const params = { issueID };

  if (subAccount) payload.subAccount = subAccount;

  return await handleRequest({
    method: HTTPMETHODS.POST,
    path: "/v0.2/accounts/issue-id-transactions",
    params,
    payload,
  });
}

/**
 * This functions returns a list of transactions by account
 * @param {string?} subAccount The subAccount of the transaction
 * @returns The response from the Chi Money API
 */
async function getAllTransactions(subAccount = null) {
  const payload = {};

  if (subAccount) payload.subAccount = subAccount;

  return handleRequest({
    method: HTTPMETHODS.POST,
    path: "/v0.2/accounts/transactions",
    payload,
  });
}

/**
 * This transaction transfers funds from wallet to receiver
 * @param {string} receiver - The receiver of the funds
 * @param {number} amount - The amount of funds
 * @param {string} wallet - The wallet to be transfered from
 * @param {string?} subAccount - The subAccount of the transaction
 * @returns The response from the Chi Money API
 */
async function accountTransfer(receiver, amount, wallet, subAccount = null) {
  // Define validation schema for input
  const schema = Joi.object({
    receiver: Joi.string().required(),
    wallet: Joi.string().required(),
    amount: Joi.number().required(),
  });

  const { value, error } = schema.validate(
    { receiver, amount, wallet },
    { abortEarly: false }
  );

  // Handle error
  if (error)
    throw new ValueError(
      "Invalid or missing function argument(s)",
      formatJoiErrors(error)
    );

  const payload = { ...value };

  if (subAccount) payload.subAccount = subAccount;

  return handleRequest({
    method: HTTPMETHODS.POST,
    payload,
    path: "/v0.2/accounts/transfer",
  });
}

/**
 * This function deletes an unpaid transaction
 * @param {string} chiRef The ID of the transaction
 * @param {string?} subAccount The subAccount of the transaction
 * @returns The response from the ChiMoney API
 */
async function deleteUnpaidTransaction(chiRef, subAccount = null) {
  if (!chiRef) throw new ValueError("transactionId is required");

  if (typeof chiRef !== "string")
    throw new TypeError("transactionId must be a string");

  const payload = { chiRef };

  if (subAccount) payload.subAccount = subAccount;

  return handleRequest({
    method: HTTPMETHODS.DELETE,
    payload,
    path: "/v0.2/accounts/delete-unpaid-transaction",
  });
}

/**
 * This function gets a transaction by ID
 * @param {string} transctionId The ID of the transaction
 * @param {string?} subAccount The subAccount of the transaction
 * @returns The response from the Chi Money API
 */
async function getTransactionByID(transctionId, subAccount = null) {
  if (!transctionId) throw new ValueError("transactionId is required");

  if (typeof transctionId !== "string")
    throw new TypeError("transactionId must be a string");

  const payload = {};
  const params = { id: transctionId };

  if (subAccount) payload.subAccount = subAccount;

  return handleRequest({
    method: HTTPMETHODS.POST,
    path: "/v0.2/accounts/transaction",
    params,
    payload,
  });
}

module.exports = {
  getAllTransactions,
  getTransactionsByIssueID,
  accountTransfer,
  deleteUnpaidTransaction,
  getTransactionByID,
};
